### PR TITLE
feat: support cert and key for LDAPS

### PR DESCRIPTION
**Pull request by flegastelois**
| Q             | A
| ------------- | ---
| Bug fix?      | yes and no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | NA

Need 3 new column in "glpi_authldap":
```
ALTER TABLE `glpi_authldaps`
   ADD COLUMN `tls_certfile` VARCHAR(255) NULL DEFAULT NULL COLLATE 'utf8_unicode_ci',
   ADD COLUMN `tls_keyfile` VARCHAR(255) NULL DEFAULT NULL COLLATE 'utf8_unicode_ci',
   ADD COLUMN `use_bind` TINYINT(1) NOT NULL DEFAULT '1';
```

These changes allow the LDAP authentication / import / directory synchronization to work with secure servers that require a certificate and a private key for connection: this explains the two added option fields (tls_certfile, tls_keyfile).

Although they are called "tls_", you must not enable TLS for the connection to work, which is why I have not conditioned their addition (ldap_set_option) to the boolean "use_tls".

Google Workspace for example:
https://support.google.com/a/answer/9190869?hl=en&ref_topic=9048334#zippy=%2Cldapsearch

```
LDAPTLS_CERT={crt_file} LDAPTLS_KEY={key_file} ldapsearch -H ldaps://ldap.google.com:636 -b dc={domain},dc={domain} '(mail={user_email})'
```

In the case of using a certificate and its private key, "ldap_bind" must not be used, this explains the added option.

I can give you production ready "certificate / key" files to add in "GLPI_CONFIG_DIR/ldap/" directory.

### CODE REVIEW

**Comment by flegastelois** [comment]
Should be interesting, to have a "Google Workspace" pre-configured schema (like "AD" or "Default") when we create a new ldap authentification server.

https://support.google.com/a/answer/9188164?hl=en&ref_topic=9048334

- Server: ldaps://ldap.google.com
- Port: 636
- ConnectionFilter: (objectClass=person)
- Use bind: no
- Login field: mail
- Synchronization fiel : entryuuid
- Use TLS: no

File: inc/authldap.class.php [hunk 1]
```diff
@@ -3506,6 +3562,8 @@ function prepareInputForAdd($input) {
          $input["rootdn_passwd"] = Toolbox::sodiumEncrypt($input["rootdn_passwd"]);
       }
 
+      $this->checkFilesExist($input);
```
Comment (cedric-anne) [inline inc/authldap.class.php:3565]:
```suggestion
      if (!$this->checkFilesExist($input)) {
         return false;
      }
```

File: inc/authldap.class.php [hunk 2]
```diff
@@ -226,6 +226,8 @@ function prepareInputForUpdate($input) {
             return false;
          };
       }
+
+      $this->checkFilesExist($input);
```
Comment (cedric-anne) [inline inc/authldap.class.php:230]:
```suggestion
      if (!$this->checkFilesExist($input)) {
         return false;
      }
```

File: inc/authldap.class.php [hunk 3]
```diff
@@ -532,6 +539,15 @@ function showFormAdvancedConfig() {
       Html::autocompletionTextField($this, "inventory_domain", ['size' => 100]);
       echo "</td></tr>";
 
+      echo "<tr class='tab_bg_1'>";
+      echo "<td>" . __('TLS Certfile') . "</td><td>";
```
Comment (cedric-anne) [inline inc/authldap.class.php:543]:
```suggestion
      echo "<td>" . __('TLS Certfile path') . "</td><td>";
```

File: inc/authldap.class.php [hunk 4]
```diff
@@ -532,6 +539,15 @@ function showFormAdvancedConfig() {
...
+      echo "<input type='text' name='tls_certfile' id='tls_certfile' value='".$this->fields["tls_certfile"]."'>";
+      echo "</td>";
+      echo "<td>" . __('TLS Keyfile') . "</td><td>";
```
Comment (cedric-anne) [inline inc/authldap.class.php:546]:
```suggestion
      echo "<td>" . __('TLS Keyfile path') . "</td><td>";
```

File: inc/authldap.class.php [hunk 5]
```diff
@@ -2628,28 +2650,47 @@ function connect() {
...
+    * @param string  $tls_keyfile   TLS KEY file name within config directory (default '')
+    * @param boolean $use_bind      do we need to do an ldap_bind? (true by default)
     *
     * @return resource link to the LDAP server : false if connection failed
     */
    static function connectToServer($host, $port, $login = "", $password = "",
-                                   $use_tls = false, $deref_options = 0) {
+                                   $use_tls = false, $deref_options = 0,
+                                   $tls_certfile = "", $tls_keyfile = "",
+                                   $use_bind = true) {
```
Comment (cedric-anne) [inline inc/authldap.class.php:2662]:
```suggestion
                                   string $tls_certfile = "", string $tls_keyfile = "",
                                   bool $use_bind = true) {
```

File: inc/authldap.class.php [hunk 6]
```diff
@@ -2628,28 +2650,47 @@ function connect() {
...
 
       $ds = @ldap_connect($host, intval($port));
       if ($ds) {
          @ldap_set_option($ds, LDAP_OPT_PROTOCOL_VERSION, 3);
          @ldap_set_option($ds, LDAP_OPT_REFERRALS, 0);
          @ldap_set_option($ds, LDAP_OPT_DEREF, $deref_options);
+
+         if (file_exists($tls_certfile)) {
```
Comment (cedric-anne) [inline inc/authldap.class.php:2670]:
```suggestion
         if (!empty($tls_certfile)) {
```

File: inc/authldap.class.php [hunk 7]
```diff
@@ -2628,28 +2650,47 @@ function connect() {
...
+            @ldap_set_option(null, LDAP_OPT_X_TLS_CERTFILE, $tls_certfile);
+         }
+
+         if ($tls_keyfile) {
```
Comment (cedric-anne) [inline inc/authldap.class.php:2674]:
```suggestion
         if (!empty($tls_keyfile)) {
```

File: inc/authldap.class.php [hunk 8]
```diff
@@ -3945,4 +4003,31 @@ public static function getUsers($values, &$results, &$limitexceeded) {
 
       return $users;
    }
+
+   public function checkFilesExist(&$input) {
+
+      if (isset($input['tls_certfile'])) {
```
Comment (cedric-anne) [inline inc/authldap.class.php:4009]: I guess we should allow removal of field value.
```suggestion
      if (isset($input['tls_certfile']) && !empty($input['tls_certfile'])) {
```

File: inc/authldap.class.php [hunk 9]
```diff
@@ -3945,4 +4003,31 @@ public static function getUsers($values, &$results, &$limitexceeded) {
...
+            Session::addMessageAfterRedirect(
+               __('TLS certificate path is incorrect'),
+               false,
+               ERROR
+            );
+            return false;
+         }
+      }
+
+      if (isset($input['tls_keyfile'])) {
```
Comment (cedric-anne) [inline inc/authldap.class.php:4021]: I guess we should allow removal of field value.
```suggestion
      if (isset($input['tls_keyfile']) && !empty($input['tls_keyfile'])) {
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

File: inc/authldap.class.php [hunk 10]
```diff
@@ -3945,4 +4007,31 @@ public static function getUsers($values, &$results, &$limitexceeded) {
...
+         $file = realpath($input['tls_keyfile']);
+         if (!file_exists($file)) {
+            Session::addMessageAfterRedirect(
+               __('TLS key file path is incorrect'),
+               false,
+               ERROR
+            );
+            return false;
+         }
+      }
```
Comment (cedric-anne) [inline inc/authldap.class.php:4035]:
```suggestion
      }

      return true;
```

File: inc/authldap.class.php [hunk 11]
```diff
@@ -3945,4 +4007,31 @@ public static function getUsers($values, &$results, &$limitexceeded) {
 
       return $users;
    }
+
+   public function checkFilesExist(&$input) {
```
Comment (cedric-anne) [inline inc/authldap.class.php:4011]:
```suggestion
   public function checkFilesExist(&$input): bool {
```

File: inc/authldap.class.php [hunk 12]
```diff
@@ -3945,4 +4007,33 @@ public static function getUsers($values, &$results, &$limitexceeded) {
 
       return $users;
    }
+
+   public function checkFilesExist(&$input): bool {
```
Comment (cedric-anne) [inline inc/authldap.class.php:4011]: Reference flag does not seems usefull.
```suggestion
   public function checkFilesExist(array $input): bool {
```

**Pull request merged**